commit 041d856e697bbba8e6c58e84d6b4c0bae4342ca4
Author: Michael Spang <spang@google.com>
Date:   Mon Aug 13 11:35:32 2012 -0400

    Add events debug log support
    
    This adds a new xinput property "Dump Debug Logs" that dumps the last
    64k input events to /var/log/evdev_input_events.dat.

diff --git a/src/Makefile.am b/src/Makefile.am
index cca1b0c..d223cb5 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -39,5 +39,6 @@ AM_CPPFLAGS =-I$(top_srcdir)/include
                                emuThird.c \
                                emuWheel.c \
                                draglock.c \
-                               apple.c
+                               apple.c \
+                               debug.c
 
diff --git a/src/debug.c b/src/debug.c
new file mode 100644
index 0000000..362e5cf
--- /dev/null
+++ b/src/debug.c
@@ -0,0 +1,101 @@
+// Copyright (c) 2012 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <evdev.h>
+#include <evdev-properties.h>
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <exevents.h>
+#include <xf86.h>
+#include <xf86Xinput.h>
+#include <X11/Xatom.h>
+
+#define XI_PROP_DUMP_DEBUG_LOG "Dump Debug Log"
+
+#define INPUT_EVENTS_LOG_FILE "/var/log/evdev_input_events.dat"
+
+static Atom dump_debug_log_prop;
+
+static void
+EvdevDumpLog(InputInfoPtr pInfo) {
+    EvdevPtr pEvdev = pInfo->private;
+    int i;
+
+    FILE *fp = fopen(INPUT_EVENTS_LOG_FILE, "w");
+    if (!fp) {
+        xf86IDrvMsg(pInfo, X_ERROR, "fopen: %s: %s\n",
+            INPUT_EVENTS_LOG_FILE, strerror(errno));
+        return;
+    }
+
+    fprintf(fp, "# device: %s\n", pInfo->name);
+
+    for (i = 0; i < DEBUG_BUF_SIZE; ++i) {
+        struct input_event *ev =
+            &pEvdev->debug_buf[(pEvdev->debug_buf_tail + i) % DEBUG_BUF_SIZE];
+        if (ev->time.tv_sec == 0 && ev->time.tv_usec == 0)
+            continue;
+        fprintf(fp, "E: %ld.%06ld %04x %04x %d\n",
+            ev->time.tv_sec,
+            ev->time.tv_usec,
+            ev->type,
+            ev->code,
+            ev->value);
+    }
+
+    fclose(fp);
+}
+
+static int
+EvdevDebugGetProperty(DeviceIntPtr dev, Atom property)
+{
+    return Success;
+}
+
+static int
+EvdevDebugSetProperty(DeviceIntPtr dev, Atom atom,
+                      XIPropertyValuePtr val, BOOL checkonly)
+{
+    InputInfoPtr pInfo = dev->public.devicePrivate;
+    CARD32 data;
+
+    if (atom == dump_debug_log_prop) {
+        if (val->type != XA_INTEGER || val->format != 32 || val->size != 1)
+            return BadMatch;
+
+        data = *(CARD32 *)val->data;
+
+        if (data != 1)
+            return BadValue;
+
+        if (!checkonly)
+            EvdevDumpLog(pInfo);
+    }
+
+    return Success;
+}
+
+void
+EvdevDebugInitProperty(DeviceIntPtr dev)
+{
+
+    dump_debug_log_prop = MakeAtom(XI_PROP_DUMP_DEBUG_LOG,
+        strlen(XI_PROP_DUMP_DEBUG_LOG), TRUE);
+    CARD32 prop_dump_debug_log_init = 0;
+
+    XIChangeDeviceProperty(dev, dump_debug_log_prop, XA_INTEGER,
+        /* format */ 32, PropModeReplace, /* size */ 1,
+        &prop_dump_debug_log_init, FALSE);
+    XISetDevicePropertyDeletable(dev, dump_debug_log_prop, FALSE);
+    XIRegisterPropertyHandler(dev, EvdevDebugSetProperty,
+        EvdevDebugGetProperty, NULL);
+} 
diff --git a/src/evdev.c b/src/evdev.c
index abf7be8..dc36e6b 100644
--- a/src/evdev.c
+++ b/src/evdev.c
@@ -1192,6 +1192,14 @@ EvdevProcessEvent(InputInfoPtr pInfo, struct input_event *ev)
             EvdevProcessSyncEvent(pInfo, ev);
             break;
     }
+
+    // Add touch events to the debug log. Keypresses are not included.
+    if (ev->type == EV_ABS || ev->type == EV_SYN ||
+        (ev->type == EV_KEY && ev->code == BTN_TOUCH)) {
+      EvdevPtr pEvdev = pInfo->private;
+      pEvdev->debug_buf[pEvdev->debug_buf_tail++] = *ev;
+      pEvdev->debug_buf_tail %= DEBUG_BUF_SIZE;
+    }
 }
 
 #undef ABS_X_VALUE
@@ -1947,6 +1955,7 @@ EvdevInit(DeviceIntPtr device)
     EvdevWheelEmuInitProperty(device);
     EvdevDragLockInitProperty(device);
     EvdevAppleInitProperty(device);
+    EvdevDebugInitProperty(device);
 
     return Success;
 }
diff --git a/src/evdev.h b/src/evdev.h
index ea2a012..c1a485f 100644
--- a/src/evdev.h
+++ b/src/evdev.h
@@ -101,6 +101,8 @@
 /* Number of longs needed to hold the given number of bits */
 #define NLONGS(x) (((x) + LONG_BITS - 1) / LONG_BITS)
 
+#define DEBUG_BUF_SIZE 65536
+
 /* Function key mode */
 enum fkeymode {
     FKEYMODE_UNKNOWN = 0,
@@ -251,6 +253,9 @@ typedef struct {
     int                     num_queue;
     EventQueueRec           queue[EVDEV_MAXQUEUE];
 
+    struct input_event debug_buf[DEBUG_BUF_SIZE];
+    size_t debug_buf_tail;
+
     enum fkeymode           fkeymode;
 
     /* Mouse scroll wheel acceleration information. */
@@ -312,4 +317,5 @@ void Evdev3BEmuInitProperty(DeviceIntPtr);
 void EvdevWheelEmuInitProperty(DeviceIntPtr);
 void EvdevDragLockInitProperty(DeviceIntPtr);
 void EvdevAppleInitProperty(DeviceIntPtr);
+void EvdevDebugInitProperty(DeviceIntPtr);
 #endif
